Create Service Node Deployment section. (#257)
Mirantis/mke-docs: content/docs/configuration/service-node-deployment.md

---

> title: Service node deployment

Reviewer: considering the comment above, I guess the page should also be renamed, but "Workload node deployment" doesn't sound as good to me, or am I tripping?

Author: Works for me.

weight: 10
---

By default, user workloads cannot run in manager nodes, to ensure that cluster
management functionality remains performant and to enhance cluster security.
You can, however, override this behavior.

{{< callout type="warning" >}}

> If for whatever reason a user deploys a malicious service that can affect

Reviewer: ```suggestion
If for whatever reason a user deploys a malicious workload that can affect
```

The term `service` comes from Docker Swarm where this was a general name for workloads, but in kubernetes, the word "service" is reserved for a special resource type which is not necessarily a workload, so I suggest that here and below we use the word `workload` instead.

the node on which it is running, that service will not be able to strike any
other nodes in the cluster or have any impact on cluster management
functionality.

{{< /callout >}}

> Restrict services deployment to Kubernetes worker nodes

Reviewer: ```suggestion
Restrict workload deployment to Kubernetes worker nodes
```

-------------------------------------------------------

By default, MKE clusters use [Kubernetes taints and tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)
to prevent user workloads from deploying to MKE manager nodes.

To view the taints, run the following command:

```
$ kubectl get nodes <mkemanager> -o json | jq -r '.spec.taints | .[]'
```

Example of system response:

```
{
  "effect": "NoSchedule",
  "key": "node-role.kubernetes.io/master"
}
```

> Allow services deployment on Kubernetes MKE manager

Reviewer: ```suggestion
Allow workload deployment on Kubernetes MKE manager
```

Reviewer: you get the idea, there are a few more places with the same edit

Author: Grabbed it up.

---------------------------------------------------

You can circumvent the protections put in place by [Kubernetes taints and
tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).

> For details, refer to [Restrict services deployment to Kubernetes worker nodes](../restrict-service-deploy-to-kube-workers).

Reviewer: there is no such page anymore

To add a toleration to the Pod specification for existing workloads:

1. Add the following toleration to the Pod specification, either through the

>    MKE web UI or using the `kubectl edit <object> <workload>` command:

Reviewer: ```suggestion
   MKE web UI or using the `kubectl edit <resource> <name>` command:
```

   ```
   tolerations:

>    - key: "node-role.kubernetes.io/master"
>      operator: "Exists"

Reviewer: I just checked how we set it for system components, and there is one more field to set
```suggestion
   - key: "node-role.kubernetes.io/master"
     operator: "Exists"
     effect: "NoSchedule"
```

   ```

2. Run the following command to confirm the successful application of the
   toleration:

   ```

>    kubectl get <object> <workload> -o json | jq -r '.spec.template.spec.tolerations | .[]'

Reviewer: this particular command doesn't work for all object types. E.g. it won't work for `kubectl get pod` but will work for `kubectl get deployment`. Maybe it would be better to just remove the command and the example and just write something like 

> Confirm the successful application of the toleration by inspecting the modified object.

   ```

Example of system response:

```

> {
> "key": "node-role.kubernetes.io/master",
> "operator": "Exists"
> }

Reviewer: ```suggestion
{
  "effect": "NoSchedule",
  "key": "node-role.kubernetes.io/master",
  "operator": "Exists"
}
```
